### PR TITLE
Remove PI gates implementation

### PR DESCRIPTION
**Pull request by maliasadi**
### Before submitting

Please complete the following checklist when submitting a PR:

- [ ] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      [`tests`](../tests) directory!

- [ ] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [ ] Ensure that the test suite passes, by running `make test`.

- [ ] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [ ] Ensure that code is properly formatted by running `make format`. 

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**
[sc-73507]

### CODE REVIEW

File: pennylane_lightning/core/src/simulators/lightning_qubit/tests/Test_StateVectorLQubit.cpp [hunk 1]
```diff
@@ -276,9 +275,7 @@ TEMPLATE_PRODUCT_TEST_CASE("StateVectorLQubit::applyMatrix with a pointer",
             const auto m = randomUnitary<PrecisionT>(re, num_wires);
 
             state_vector_1.applyMatrix(m, wires);
-
-            Gates::GateImplementationsPI::applyMultiQubitOp<PrecisionT>(
-                state_vector_2.getData(), num_qubits, m.data(), wires, false);
+            state_vector_2.applyMatrix(m, wires);
```
Comment (multiphaseCFD) [inline pennylane_lightning/core/src/simulators/lightning_qubit/tests/Test_StateVectorLQubit.cpp:278]: Could we just remove this test? The results will match always.

File: pennylane_lightning/core/src/utils/TestKernels.hpp [hunk 2]
```diff
@@ -22,9 +22,7 @@
 #include "TypeList.hpp"
 
 #include "cpu_kernels/GateImplementationsLM.hpp"
-#include "cpu_kernels/GateImplementationsPI.hpp"
 
 using TestKernels = Pennylane::Util::TypeList<
-    Pennylane::LightningQubit::Gates::GateImplementationsPI,
     Pennylane::LightningQubit::Gates::GateImplementationsLM, void>;
 #endif
```
Comment (multiphaseCFD) [inline pennylane_lightning/core/src/utils/TestKernels.hpp:28]: What about adding a blank line at the end of this file?

File: pennylane_lightning/core/src/simulators/lightning_qubit/gates/tests/Test_KernelMap.cpp [hunk 3]
```diff
@@ -35,7 +35,7 @@ using Pennylane::Util::LightningException;
 
 TEST_CASE("Test PriorityDispatchSet", "[PriorityDispatchSet]") {
     auto pds = PriorityDispatchSet();
-    pds.emplace(Pennylane::Gates::KernelType::PI, 10U,
+    pds.emplace(Pennylane::Gates::KernelType::LM, 10U,
```
Comment (multiphaseCFD) [inline pennylane_lightning/core/src/simulators/lightning_qubit/gates/tests/Test_KernelMap.cpp:38]: I guess we can just remove it

File: pennylane_lightning/core/src/simulators/lightning_qubit/gates/cpu_kernels/GateImplementationsPI.hpp [hunk 4]
```diff
@@ -1,1394 +0,0 @@
...
-#define _USE_MATH_DEFINES // for C++
-#endif
-/// @endcond
-
-#include <bit>
-#include <complex>
-#include <vector>
-
-#include "BitUtil.hpp"
-#include "GateIndices.hpp"
```
Comment (multiphaseCFD) [inline pennylane_lightning/core/src/simulators/lightning_qubit/gates/cpu_kernels/GateImplementationsPI.hpp:32]: Do we want to keep `GateIndices.hpp`?

**Review by multiphaseCFD — COMMENTED** [review]
Thanks @maliasadi ! LGTM.

**Review by AmintorDusko — APPROVED** [review]
WOW beautiful!

**Comment by AmintorDusko** [comment]
Please test also the wheels.

File: doc/lightning_qubit/development/avx_kernels/build_system.rst [hunk 5]
```diff
@@ -8,11 +8,10 @@ when the library is loaded, and it is used at runtime when it is the most suitab
 To support AVX2 and AVX512 kernels, we always compile those kernels if the target system is UNIX on x86-64.
 Specifically, we made separate C++ files for AVX2 and AVX512 kernels and build them as a static library with the corresponding compile options. This is handled by CMake. One can check ``gates/CMakeLists.txt`` file for details.
 
-One caveat is that we want to make sure that default kernels (``KernelType::PI`` and ``KernelType::LM``) are only instantiated once with specific compiler flags during the compile process.
+One caveat is that we want to make sure that default ``KernelType::LM`` kernels are only instantiated once with specific compiler flags during the compile process.
```
Comment (LuisAlfredoNu) [inline doc/lightning_qubit/development/avx_kernels/build_system.rst:11]:
```suggestion
One caveat is that we want to ensure that default ``KernelType::LM`` kernels are only instantiated once with specific compiler flags during the compile process.
```

File: doc/lightning_qubit/development/avx_kernels/build_system.rst [hunk 6]
```diff
@@ -8,11 +8,10 @@ when the library is loaded, and it is used at runtime when it is the most suitab
...
 This is important as the linker sometimes cannot choose the right instantiation when there are multiple instantiations of the same template class.
 This problem does not arise when all instantiations are compiled with the same options, but with the AVX2/512 kernels, we use different compile options for each translation unit. We solve this problem by adding explicit instantiation declarations in the header files for these kernels
-(:ref:`file_pennylane_lightning_core_src_simulators_lightning_qubit_gates_cpu_kernels_GateImplementationsLM.hpp` and
-:ref:`file_pennylane_lightning_core_src_simulators_lightning_qubit_gates_cpu_kernels_GateImplementationsPI.hpp`)
+(:ref:`file_pennylane_lightning_core_src_simulators_lightning_qubit_gates_cpu_kernels_GateImplementationsLM.hpp`)
```
Comment (LuisAlfredoNu) [inline doc/lightning_qubit/development/avx_kernels/build_system.rst:14]: If you click the link, the website does not render correctly. 
image

**Review by LuisAlfredoNu — COMMENTED** [review]
Looks very nice, this will speed up the compilation time. Thank you @maliasadi 
Just a couple of comments

File: pennylane_lightning/core/src/simulators/lightning_qubit/gates/tests/Test_KernelMap.cpp (hunk 3 above)
Comment (maliasadi) [reply pennylane_lightning/core/src/simulators/lightning_qubit/gates/tests/Test_KernelMap.cpp:38]: Let's keep it to ensure that the priority dispatcher always works as expected even on Windows with one single gate match.

File: pennylane_lightning/core/src/simulators/lightning_qubit/tests/Test_StateVectorLQubit.cpp (hunk 1 above)
Comment (maliasadi) [reply pennylane_lightning/core/src/simulators/lightning_qubit/tests/Test_StateVectorLQubit.cpp:278]: Yeah, let's remove it!

File: pennylane_lightning/core/src/utils/TestKernels.hpp (hunk 2 above)
Comment (maliasadi) [reply pennylane_lightning/core/src/utils/TestKernels.hpp:28]: Thanks!

File: doc/lightning_qubit/development/avx_kernels/build_system.rst (hunk 6 above)
Comment (maliasadi) [reply doc/lightning_qubit/development/avx_kernels/build_system.rst:14]: This issue has been there as long as I could traverse over the history of API docs :(

Comment (maliasadi) [reply doc/lightning_qubit/development/avx_kernels/build_system.rst:14]: This issue has existed for as long as I can trace back through the history of the API docs :(

File: pennylane_lightning/core/src/simulators/lightning_qubit/gates/cpu_kernels/GateImplementationsPI.hpp (hunk 4 above)
Comment (maliasadi) [reply pennylane_lightning/core/src/simulators/lightning_qubit/gates/cpu_kernels/GateImplementationsPI.hpp:32]: Why should we remove it? :thinking:

**Review by LuisAlfredoNu — APPROVED** [review]
Thank you @maliasadi 
Nice work! 

Just one question. Why do we remove the `GateImplementationsPI`?

**Comment by LuisAlfredoNu** [comment]
Another question. Do the other devices (LK and LGPU) have the `GateImplementationsPI`? :thinking:

File: pennylane_lightning/core/src/simulators/lightning_qubit/gates/cpu_kernels/GateImplementationsPI.hpp (hunk 4 above)
Comment (multiphaseCFD) [reply pennylane_lightning/core/src/simulators/lightning_qubit/gates/cpu_kernels/GateImplementationsPI.hpp:32]: I thought  only `GateImplementationsPI.hpp` requires `GateIndices.hpp`. Never mind.

**Review by multiphaseCFD — APPROVED** [review]
Nice job! Thanks @maliasadi.

**Comment by maliasadi** [comment]
Thanks for your review @LuisAlfredoNu!

> Just one question. Why do we remove the GateImplementationsPI?

The PI gates were the first implementation of gate kernels in `"lightning.qubit"` using pre-computed indices, prior to the development of LM (less memory) and AVX kernels. We shouldn't use these kernels, as LM kernels are faster than PI gates for all 'supported' gates and are now our base kernels across different platforms or they should be on macOS ;) thus, we can deprecate them without any performance penalty in favour of reducing compilation time and ensuring that we only rely on LM kernels in the dynamic dispatcher across all platforms.

> Another question. Do the other devices (LK and LGPU) have the GateImplementationsPI? 🤔

No, LK follows LM kernel design and LGPU is backed by `custatevec`.

**Pull request merged**